include/psa/crypto_struct.h

```diff
@@ -89,35 +89,10 @@ typedef struct {
 
 struct psa_hash_operation_s
 {
-    psa_algorithm_t alg;
-    union
-    {
-        unsigned dummy; /* Make the union non-empty even with no supported algorithms. */
-#if defined(MBEDTLS_MD2_C)
-        mbedtls_md2_context md2;
-#endif
-#if defined(MBEDTLS_MD4_C)
-        mbedtls_md4_context md4;
-#endif
-#if defined(MBEDTLS_MD5_C)
-        mbedtls_md5_context md5;
-#endif
-#if defined(MBEDTLS_RIPEMD160_C)
-        mbedtls_ripemd160_context ripemd160;
-#endif
-#if defined(MBEDTLS_SHA1_C)
-        mbedtls_sha1_context sha1;
-#endif
-#if defined(MBEDTLS_SHA256_C)
-        mbedtls_sha256_context sha256;
-#endif
-#if defined(MBEDTLS_SHA512_C)
-        mbedtls_sha512_context sha512;
-#endif
-    } ctx;
+    psa_operation_driver_context_t ctx;
```

Comment:
I'd like to make a case for rather applying this format and re-organising cipher.

The reasoning is stack usage. If we keep the gigantic union, it will expand to the size of the largest context supported in software, regardless of which algorithm you actually are using at that time. If you only need SHA-512 sporadically, it doesn't make sense to stack-allocate 212 bytes (on a 32-bit system) for each hash operation when the rest of them can be hardware-accelerated with potentially as little as 12 bytes.

Can you elaborate on what exactly the mbed TLS implementation of PSA Crypto cares about with regards to memory allocation for operation contexts? The way I see it, we are doing memory allocation for driver operation contexts already. When a software implementation behaves like a pseudo-driver, I don't see why it would then need special treatment for its context structures, allocation-wise. Having dynamic allocation for *all* contexts passing through the driver-wrapper layer makes it in addition more homogenous to implement.

Reply:
I understand your point. For sure there is a tradeoff here. @gilles-peskine-arm it would be great to have your opinion on this.

> Can you elaborate on what exactly the mbed TLS implementation of PSA Crypto cares about with regards to memory allocation for operation contexts?
I think there is a long term goal to try to not do any dynamic memory allocation in the Mbed TLS PSA crypto implementation for robustness/security reasons.

> When a software implementation behaves like a pseudo-driver, I don't see why it would then need special treatment for its context structures, allocation-wise. Having dynamic allocation for all contexts passing through the driver-wrapper layer makes it in addition more homogeneous to implement.

I didn't think about it thoroughly but the other way around when compiling the library for a given set of drivers, we can imagine that the driver context types are added to the operation structure union and thus there is no dynamic allocation neither for software implementation nor for drivers.

Reply:
There is significant demand for eliminating heap allocation in the crypto part of Mbed TLS. A design goal of the PSA crypto API was to make it reasonable to implement without heap allocation. In Mbed TLS, we're far from not needing dynamic allocation in asymmetric crypto. On the other hand, for symmetric crypto, the low-level layers (`aes.c`, `sha512.c`, etc.) don't do any dynamic allocation, and PSA code should not do any dynamic allocation for individual operations outside the parts that are still using the classic `cipher` module.

I didn't look at the details here, but if `psa_hash_setup` starts calling malloc, that's a blocker. The only allocation should be the allocation of the context structure which is done by the application.

Yes, as a consequence, the size of the operation structure is the largest needed size among supported algorithms. This is usually not a big concern because constrained platforms where stack or heap size is at a premium usually don't implement algorithms with large memory consumption.

Reply:
If that's the design goal, then all previous driver work needs to be revisited as well, since all plugged-in drivers with multipart support would need to contribute their respective operation context sizes to the overarching struct union in `crypto_struct.h`. I don't see an immediate way of achieving that.

The obvious way to go about it would be to make `psa_driver_wrappers.h` be autogenerated as well, and have it contain includes for all drivers, as well as define the union of all context structures.

At this point, if it's a stated design goal to not dynamically allocate operation contexts, I'm wondering why the current driver wrapper implementation even got put in place with its dynamic allocation of driver contexts.

Reply:
For the time being in the driver wrapper there are dynamic allocations only for the driver context of cipher multi-part operations. That's ok I think for the time being to expand this to the other multi-part driver operation contexts but better to not go beyond that as that not really the way we want it to be eventually. Then, when implementing the generation of the wrapper, we can revisit that and try and remove the dynamic allocations for the multi-part driver operation contexts. I have just amended #3342 to not forget about it. Does it seem a reasonable plan?

Reply:
I don't think I'm on board with that. It's extremely wasteful to have *both* those 212 bytes always on the stack and then allocate an additional 212 bytes (for arguments' sake) on the heap when the wrapper figures out it can use a driver.

Those allocations should be aliased, either through always putting them on the heap (which you apparently don't want) or by making the context size for all compiled-in drivers visible to `crypto_struct.h`. Unless someone sees a third solution which avoids needlessly allocating large swaths of memory?

Reply:
Why an additional 212 bytes on the heap ? The driver context should be only of the size needed by the driver.

Reply:
Yes, it will only be the size needed by the driver. But like I said, for argument's sake, let's assume that the driver needs to store the same amount of context as the software implementation would, i.e. 212.

Reply:
> by making the context size for all compiled-in drivers visible to `crypto_struct.h`. 

That's the plan.

Reply:
Okay, so if that's the plan, then all work on the driver dispatch should adhere to the plan, right? Why wasn't this an issue 6 months ago on #3644 then?

Frustration aside: how do you see this progressing? By updating this PR to declare `psa_driver_wrappers.h` as future autogenerated file, move the #includes for the different drivers up into the header file, and make `crypto_struct.h` depend on the driver headers?

Reply:
I've proposed a way forward above where we keep allocating the driver contexts while we don't have the wrapper generation. Definitely not a possible option on your side? On first sight, the two times 212 bytes of memory seem a bit of an edge case to me where you want to do the most memory expensive hash operation both in software and in hardware.

Reply:
#3644 was about symmetric ciphers, which currently go through the `mbedtls_cipher` abstraction which calls calloc, so it wasn't a regression. This PR is about hashes, which are already malloc-free.

When the driver wrapper generation code is available, `crypto_struct.h` will include a header created by the generation code. In the meantime, there needs to be public a header file that defines the driver context types, which driver vendors will have to edit manually.

Reply:
> I've proposed a way forward above where we keep allocating the driver contexts while we don't have the wrapper generation. Definitely not a possible option on your side?

That's fine by me as long as users who don't use drivers aren't affected. If you use the software implementation, calling `psa_hash_setup` must not start causing a heap allocation.

Reply:
> Definitely not a possible option on your side?

Not on this scale. With the cipher operation PSA context the overhead was excusable because the cipher context itself contains dynamically-allocated contexts as Gilles pointed out (though it still isn't small by any account). With hash, where having SHA-512 implemented in fallback is a common usecase, it's too much.

Back to the drawing board, then. Would somebody care to actually document this (PSA Core wants to actively get rid of dynamic allocations) somewhere? It's new to me, at least.

Reply:
> Would somebody care to actually document this (PSA Core wants to actively get rid of dynamic allocations) somewhere?

I've started compiling a list of such things at https://github.com/ARMmbed/mbedtls-docs/pull/5